cli/internal/fs/package_json.go

```diff
@@ -120,6 +122,24 @@ func MarshalPackageJSON(pkgJSON *PackageJSON) ([]byte, error) {
 	return b.Bytes(), nil
 }
 
+// SetExternalDeps sets TransitiveDeps and populates ExternalDepsHash
+func (p *PackageJSON) SetExternalDeps(externalDeps mapset.Set) error {
+	p.Mu.Lock()
+	defer p.Mu.Unlock()
+	p.TransitiveDeps = make([]lockfile.Package, 0, externalDeps.Cardinality())
+	for _, dependency := range externalDeps.ToSlice() {
+		dependency := dependency.(lockfile.Package)
+		p.TransitiveDeps = append(p.TransitiveDeps, dependency)
+	}
+	sort.Sort(lockfile.ByKey(p.TransitiveDeps))
```

Comment:
Quick history of the `TransitiveDeps` field:
 - Initially this was a string of the format `{packageName}@{version}`, worked when we just had yarn1 support, but for pnpm these two pieces of information weren't enough. So we moved to using lockfile keys instead.
 - I moved from the `{lockfileKey}@{version}` to the `Package` struct (a mistake in hindsight), just because the string wasn't useful unless you correctly split it into two parts and I was afraid of someone (myself) doing `strings.Split("/@babel/types@1.2.3@1.2.3", "@")` to try to get the version and end up with garbage
 - Most recently there was an issue where pnpm package aliases where we weren't correctly merging lockfile keys. We basically had an issue of the same package being listed as `Package{Key: "/foo/1.2.3", Version: "1.2.3"}` and another as `Package{Key: "/foo/1.2.3", Version: "npm:foo@1.2.3"}` (See #4555 for a more complete description).